hal/src/thumbv7em/icm.rs

```diff
@@ -0,0 +1,1612 @@
+    /// This might result in unexpected results for example
+    /// when sequentially trying to determine which interrupt triggered.
+    ///
+    /// This is an alternative, return all the data from the register
+    /// and parse later with the designated `get_[name]_int` functions.
+    #[inline]
+    pub fn get_interrupt_status(&mut self) -> IcmInterrupt {
+        let interrupt_vector = self.token.isr().read().bits();
+        IcmInterrupt { interrupt_vector }
+    }
+    /// Trigger recalculation of memory monitor region specified
+    /// by the bitmask:
+    /// 0b0001 = region0
+    /// 0b0010 = region1
+    /// 0b0100 = region2
+    /// 0b1000 = region3
+    /// ...
+    /// 0b1111 = all regions
+    #[inline]
+    pub fn trigger_rehash(&self, bitmask: u8) {
+        self.token
+            .ctrl()
+            .write(|w| unsafe { w.rehash().bits(bitmask) });
+    }
+}
+
+/// Trait providing numerical identifier and
+/// offset for each ICM Region
+///
+/// ICM supports 4 memory regions
+pub trait IcmRegionNum: Sealed {
+    /// Numerical ID of the memory region
+    const NUM: usize;
+    /// Memory offset
+    const OFFSET: u32;
+}
+
+/// ICM Region 0
+pub enum Region0 {}
+impl Sealed for Region0 {}
+impl IcmRegionNum for Region0 {
+    const NUM: usize = 0;
+    const OFFSET: u32 = 0;
+}
+
+/// ICM Region 1
+pub enum Region1 {}
+impl Sealed for Region1 {}
+impl IcmRegionNum for Region1 {
+    const NUM: usize = 1;
+    const OFFSET: u32 = 0x10;
+}
+
+/// ICM Region 2
+pub enum Region2 {}
+impl Sealed for Region2 {}
+impl IcmRegionNum for Region2 {
+    const NUM: usize = 2;
+    const OFFSET: u32 = 0x10 * 2;
+}
+
+/// ICM Region 3
+pub enum Region3 {}
+impl Sealed for Region3 {}
+impl IcmRegionNum for Region3 {
+    const NUM: usize = 3;
+    const OFFSET: u32 = 0x10 * 3;
+}
+
+/// Functions required by [`IcmRegionDesc`]
+///
+/// Both Main List descriptors and Secondary List descriptors
+pub trait IcmRegionDescT {
+    /// Set the [`Raddr`] start of the region memory region
+    fn set_raddr<T>(&mut self, addr: *const T);
+    /// Set the specific region configuration
+    fn set_rcfg(&mut self, cfg: Rcfg);
+    /// Set the link to next region descriptor
+    fn set_rnext(&mut self, next: Rnext);
+    /// Reset Rcfg to default values
+    fn reset_rcfg_to_default(&mut self);
+}
+
+/// Helper for creating the Region Descriptor structure
+///
+/// It is also possible to construct the IcmRegion manually,
+/// but then care has to be taken to point `rnext` to the appropriate
+/// place in memory, here the hardware assumption of 0x10
+/// offset to the next descriptor is ensured.
+///
+/// Follows C-structure conventions and is 64-byte aligned
+///
+/// >**Important**
+/// >
+/// >ICM engine will **read** wherever this
+/// >is instantiated in memory, based on the [`Icm::set_dscr_addr()`]
+/// >so the user must ensure that this variable lives long enough or is
+/// >static
+#[repr(C)]
+#[repr(align(64))]
+pub struct IcmRegion {
```

Comment:
As mentioned above, I would rename this to `icm::Regions`, with an `s`.

Reply:
Adopted